src/libraries/System.Runtime.Numerics/src/System/Numerics/BigNumber.cs

```diff
@@ -380,79 +382,254 @@ internal static BigInteger ParseBigInteger(ReadOnlySpan<char> value, NumberStyle
             if (!TryValidateParseStyleInteger(style, out e))
                 throw e;
 
-            BigInteger result = BigInteger.Zero;
-            if (!TryParseBigInteger(value, style, info, out result))
+            if (!TryParseBigInteger(value, style, info, out BigInteger result))
             {
                 throw new FormatException(SR.Overflow_ParseBigInteger);
             }
             return result;
         }
 
-        private static unsafe bool HexNumberToBigInteger(ref BigNumberBuffer number, ref BigInteger value)
+        private static bool HexNumberToBigInteger(ref BigNumberBuffer number, out BigInteger result)
         {
             if (number.digits == null || number.digits.Length == 0)
+            {
+                result = default;
                 return false;
+            }
+
+            const int DigitsPerBlock = 8;
+
+            int totalDigitCount = number.digits.Length - 1;   // Ignore trailing '\0'
+            int blockCount, partialDigitCount;
 
-            int len = number.digits.Length - 1; // Ignore trailing '\0'
-            byte[] bits = new byte[(len / 2) + (len % 2)];
+            if ((uint)totalDigitCount % DigitsPerBlock == 0)
+            {
+                blockCount = totalDigitCount / DigitsPerBlock;
+                partialDigitCount = 0;
+            }
+            else
+            {
+                blockCount = Math.DivRem(totalDigitCount, DigitsPerBlock, out int mod) + 1;
+                partialDigitCount = DigitsPerBlock - mod;
+            }
+
+            bool isNegative = HexConverter.FromChar(number.digits[0]) >= 8;
+            uint partialValue = (isNegative && partialDigitCount > 0) ? 0xFFFFFFFFu : 0;
+
+            int[]? arrayFromPool = null;
 
-            bool shift = false;
-            bool isNegative = false;
-            int bitIndex = 0;
+            Span<int> bitsBuffer = (blockCount <= BigInteger.StackallocUInt32Limit)
+                ? stackalloc int[blockCount]
+                : (arrayFromPool = ArrayPool<int>.Shared.Rent(blockCount)).AsSpan(0, blockCount);
```

Comment:
I'm not a huge fan of switching this to be signed as it can be more error prone with certain shifting or arithmetic.
We could use `MemoryMarshal.Cast` to reinterpret the array as a `Span<uint>`

```suggestion
            Span<uint> bitsBuffer = (blockCount <= StackallocUInt32Limit)
                ? stackalloc uint[blockCount]
                : MemoryMarshal.Cast<int, uint>((arrayFromPool = ArrayPool<int>.Shared.Rent(blockCount)).AsSpan(0, blockCount));
```